lib/axiom/fs/stream/memory_stream_buffer.js

```diff
@@ -24,116 +24,35 @@ var ReadableStream;
 /** @typedef WritableStream$$module$axiom$fs$stream$writable_stream */
 var WritableStream;
 
-/** @typedef function():void */
-var EventCallback;
-
-/**
- * Event value + associated callback
- * @constructor
- * @param {!*} value
- * @param {!EventCallback} callback
- */
-var EventWithCallback = function(value, callback) {
-  this.value = value;
-  this.callback = callback;
-};
-
 /**
  * Implementation of streams over an in-memory array.
  *
- * @constructor
+ * @constructor @extends {StreamBuffer}
  */
 export var MemoryStreamBuffer = function() {
-  /** @private @type {boolean} */
-  this.paused_ = true;
-  /** @const @private @type {Queue} */
-  this.events_ = new Queue();
-  /** @type {!AxiomEvent} */
-  this.onData = new AxiomEvent();
-  /** @type {!AxiomEvent} */
-  this.onReadable = new AxiomEvent();
+  StreamBuffer.call(this);
+
   /** @const @type {!ReadableStream} */
   this.readableStream = new ReadableMemoryStream(this);
   /** @const @type {!WritableStream} */
   this.writableStream = new WritableMemoryStream(this);
-};
-
-export default MemoryStreamBuffer;
 
-/**
- * @return {void}
- */
-MemoryStreamBuffer.prototype.pause = function() {
-  this.paused_ = true;
-};
-
-/**
- * @return {void}
- */
-MemoryStreamBuffer.prototype.resume = function() {
-  this.paused_ = false;
-  this.flushEvents_();
+  this.writableStream.onFinish.listenOnce(function() {
+    // This will fire an event chain onFlush -> this.readableStream.onEnd.
+    StreamBuffer.prototype.flush.call(this);
```

Comment:
Because `this.flush()` would call `writableStream.end()`, which would fire `writableStream.onFinish`, which would loop back into this handler in an infinite loop. `this.flush()` is designated for external clients.